lightning/src/ln/channelmanager.rs

```diff
@@ -698,6 +698,20 @@ macro_rules! maybe_break_monitor_err {
 	}
 }
 
+/// When claiming a payment after receiving a preimage, there's a rare case
+/// where the channel hits the chain before the `ChannelMonitor` can be updated
+/// with knowledge of the preimage. Thus, monitor information needs to be passed
+/// to the `ChannelManager` on receipt of a preimage so it's capable of updating
+/// the relevant `ChannelMonitor` after the channel's data has been removed from
+/// the `ChannelManager`'s `ChannelHolder`.
```

Comment:
See my point above, but I fear we need to change our channel force-close model to conserve a channel reference until all its HTLC are closed. When we learn a preimage from the downstream offchain channel we need access to both the funding outpoint and monitor update id. 

An alternative would be to add upstream's funding outpoint/monitor_update_id in PreviousHopData but still you would have to update monitor_update_id of all forward HTLCs, this sounds gross.